utils/filecache/cache.go

```diff
@@ -0,0 +1,227 @@
+package filecache
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/ARM-software/golang-utils/utils/commonerrors"
+	fs "github.com/ARM-software/golang-utils/utils/filesystem"
+	"github.com/ARM-software/golang-utils/utils/parallelisation"
+)
+
+type CacheEntry struct {
+	path       string
+	expiration time.Time
+	ttl        time.Duration
+}
+
+type Cache struct {
+	entries       map[string]*CacheEntry
+	fs            fs.FS
+	cfg           *FileCacheConfig
+	gcCancelStore *parallelisation.CancelFunctionStore
+	entryProvider IEntryProvider
+	mu            sync.RWMutex
+	closed        bool
+}
+
+func (c *Cache) gc(ctx context.Context, _ time.Time) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	now := time.Now()
+	for id, entry := range c.entries {
+		if now.After(entry.expiration) {
+			if err := c.fs.RemoveWithContext(ctx, entry.path); err == nil {
+				delete(c.entries, id)
+			}
+		}
+	}
+}
+
+func (c *Cache) checkIfClosed() error {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	if c.closed {
+		return commonerrors.New(commonerrors.ErrConflict, "cache is closed")
+	}
+
+	return nil
+}
+
+func (c *Cache) Has(ctx context.Context, key string) (bool, error) {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	if err := c.checkIfClosed(); err != nil {
+		return false, err
+	}
+
+	_, exists := c.entries[key]
+	return exists, nil
```

Comment:
try to only lock for the shortest time possible, only where it is absolutely needed (i.e. map access itself) and don't lock around another function that locks.

```suggestion
func (c *Cache) Has(ctx context.Context, key string) (bool, error) {
	if err := c.checkIfClosed(); err != nil {
		return false, err
	}

	c.mu.RLock()
	defer c.mu.RUnlock()
	
	_, exists := c.entries[key]
	return exists, nil
```